### ISSUE TITLE
Files included in CentOS RPM interfere with CNI operation

### ISSUE THREAD

**Issue by matthias50** [cri-o/cri-o#2885]
**Description**

When I deploy Calico CNI to together with crio, the overlay network doesn't work at all.

**Steps to reproduce the issue:**
1. Install crio on nodes
2. Deploy calico

**Describe the results you received:**

crio itself seems to be working. The nodes comes and pods get created there. Calico daemonset up and running, but the networking does work. When I run `ifconfig -a`, Instead of seeing a bunch of devices called `calixxxxxxxxxxx` like I would expect, I get devices called `vethxxxxxxxxxxx`. 

**Describe the results you expected:**

Calico CNI networking functions normally.

**Additional information you deem important (e.g. issue happens only occasionally):**

After digging into this a bit, it seems that the issue is that the crio CentOS rpm deploys these two files:  `/etc/cni/net.d/100-crio-bridge.conf` and `/etc/cni/net.d/200-loopback.conf`. 

```
$ rpm -qf  /etc/cni/net.d/100-crio-bridge.conf
cri-o-1.14.9-1.el7.x86_64
$ rpm -qf  /etc/cni/net.d/200-loopback.conf
cri-o-1.14.9-1.el7.x86_64
```

These files create a misconfiguration which interferes with Calico, and I suspect, any CNI. If I delete these files after installing the RPM and before starting Calico and kubelet, Calico functions normally.

**Output of `crio --version`:**

```
crio version 1.14.9-1.el7
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

```
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

**Comment by ezbin** [comment]
I've also encountered this issue on CentOS 7.  I came to the same work around, with one additional step.

The `crio.conf` file generated by kubespray will only set the `plugin_dir` networking parameter to `"/usr/libexec/cni"`.  If you delete the `100-crio-bridge.conf` file with this setup, then cri-o will fail to start with an error about no default network.

This happens because all of the calico plugins are stored in `/opt/cni/bin`, which means that the calico cni configuration files in `/etc/cni/net.d/` will fail to load.

The `crio.conf` template will only include `/opt/cni/bin` if the OS is ClearLinux.  If you update the conditional to also include RedHat, then everything will work.

**Comment by haircommander** [comment]
@matthias50 wow, totally missed this. I assume this is still an issue, but I don't know if there's much we can do. We need to ship a set of configurations for users that don't want to set their own, so the installation works OOTB. Is the work around removing those configurations and putting in calico good enough? (I'd guess it is, because this issue is nearly a year old with 0 work having been done for it :sweat_smile: )

**Comment by matthias50** [comment]
The work-around of removing the files does allow crio + Calico to function. However, from my perspective, this very much did *not* work OOTB. Rather, I had to to do a lot of digging in order to figure out why Calico was blowing up spectacularly with crio when it worked fine with docker.  I guess the "box" you are referring to is crio without a CNI, but too me, that isn't a realistic use case.

**Comment by bart0sh** [comment]
@haircommander I also experienced this issue recently. I've tried 3 CNI plugins: calico, weave and flannel and had networking messed up with all of them. The issues were hard to debug.
Can you explain why CRI-O installs 100-crio-bridge.conf and which use cases does it solve? If those use cases are not often used it would make sense to remove CNI configuration files from the CRI-O packages in my opinion.

**Comment by haircommander** [comment]
> Can you explain why CRI-O installs 100-crio-bridge.conf and which use cases does it solve?

there are cases where users don't want the complex cni plugins and just want simple host<->container communication, and the bridge plugin satisfies that.

What files do calico, weave and flannel create in that directory? Maybe we can make it so that the bridge plugin has lower priority than those and they're chosen instead. That was the intention, but I see it doesn't seem to be working

**Comment by bart0sh** [comment]
@haircommander 
weave creates 10-weave.conflist, calico doesn't create anything in the /etc/cni/net.d and flannel creates 10-flannel.conflist

However, the problem as I can see it is that k8s control plane is usually setup without installing any CNI plugin. At least this is the workflow kubeadm suggests to use. So, if runtime (CRI-O in this case) enables CNI it breaks that workflow. As a result coredns pods are assigned IPs provided by already active CNI plugin(bridge). When user installs another plugin it doesn't change IP addresses of the coredns pods. When new pods are created they can get IP addresses from another CNI plugin and wouldn't be able to connect to coredns pods. That can create name resolution and other networking issues.

Here is an example of the cluster setup done with kubeadm:
1. Install CRI-O and run it (it activates CNI bridge plugin)
2. kubeadm init bootstraps a cluster and suggests to install CNI:
...
Your Kubernetes control-plane has initialized successfully!
...
You should now deploy a pod network to the cluster.
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:
  https://kubernetes.io/docs/concepts/cluster-administration/addons/
...
3. It's already late as coredns pods already got their IPs from bridge CNI:
```
$ k get pods -A -o wide
NAMESPACE     NAME                            READY   STATUS    RESTARTS   AGE     IP              NODE    NOMINATED NODE   READINESS GATES
kube-system   coredns-78fcd69978-g6dgz        1/1     Running   0          2m36s   10.85.0.3       devel   <none>           <none>
kube-system   coredns-78fcd69978-lwtwv        1/1     Running   0          2m36s   10.85.0.2       devel   <none>           <none>
kube-system   etcd-devel                      1/1     Running   5          2m36s   10.237.72.187   devel   <none>           <none>
kube-system   kube-apiserver-devel            1/1     Running   5          2m36s   10.237.72.187   devel   <none>           <none>
kube-system   kube-controller-manager-devel   1/1     Running   5          2m36s   10.237.72.187   devel   <none>           <none>
kube-system   kube-proxy-srjdc                1/1     Running   0          2m36s   10.237.72.187   devel   <none>           <none>
kube-system   kube-scheduler-devel            1/1     Running   5          2m36s   10.237.72.187   devel   <none>           <none>
```

Does this makes sense for you?

**Comment by haircommander** [comment]
ah yes this makes sense. I suppose the solution would be to separate the CNI plugin installation from CRI-O, and have the CNI plugin installation optional. We would also need to update documentation to make sure users see how best to install the bridge plugin if they're looking for that.

out of curiosity, what operating system are you installing on @bart0sh ?

**Comment by matthias50** [comment]
> What files do calico, weave and flannel create in that directory?

In case it helps, with our usage of Calico, we have /etc/cni/net.d/10-calico.conflist

**Comment by bart0sh** [comment]
@haircommander I reproduced it at least on Ubuntu 18.04, Ubuntu 20.04 and OpenSUSE Leap 15.2

Yes, making CNI plugin installation optional and disabled by default would help, I believe.
If you can briefly explain how to do it I'll try to come up with a PR.

**Comment by haircommander** [comment]
Okay I have discussed this with @lsm5 . We can set the containernetworking-plugins as "Suggests" instead of "Recommends" in the spec file. I think we should wait until 1.24 to do so, to simplify documentation changes and not confuse users. Is that acceptible @bart0sh @matthias50 ?

**Comment by bart0sh** [comment]
@haircommander 
> Is that acceptible

Absolutely

**Comment by volker-raschek** [comment]
Hi @bart0sh,

>  I also experienced this issue recently. I've tried 3 CNI plugins: calico, weave and flannel and had networking messed up with all of them. 

I have exactly this problem with all of this CNI plugins on my Arch Linux cluster and I can commit that all of my coredns pods get an IP-Adresse of `10.85.0.0` subnet. I was very irritated, because I specified `10.244.0.0/16` as pod-cidr. DNS queries within the cluster do not work. 

Is there a workaround to solve the problem? I have tried to remove the `/etc/cni/net.d/100-crio-bridge.conf` and  `/etc/cni/net.d/199-crio-loopback.conf`, but unfortunately the coredns pods do not start after rebooting the nodes with weave as CNI plugin.

**Comment by bart0sh** [comment]
@volker-raschek I'm using the following workaround:
 - install cri-o
 - delete everything from /etc/cni/net.d
 - setup node with kubeadm
 - install cni plugin (wave, flannel, calico, etc)
 
This way you can ensure you have only one cni plugin running and all pods will get addresses from it

**Comment by bart0sh** [comment]
@haircommander
> We can set the containernetworking-plugins as "Suggests" instead of "Recommends" in the spec file.

Does this mean that it will work only for rpm distros? Can you point me out to the deb packages? Are those hosted by this project or done by distros themselves?

**Comment by haircommander** [comment]
> Does this mean that it will work only for rpm distros? Can you point me out to the deb packages? Are those hosted by this project or done by distros themselves?

they're also packaged by me. I'll have to investigate a bit. it currently looks like we only `Suggests:` for deb packages, so maybe we need to update the install.md to skip installing instead

**Comment by bart0sh** [comment]
@haircommander btw, containernetworking-plugins doesn't contain CNI configuration files on Ubuntu. Are you planning to move them there?

**Comment by haircommander** [comment]
hm maybe I've confused myself with the packaging structure. it's possible the ones we ship are wrapped up in our containers-common package...

**Comment by volker-raschek** [comment]
Hi  @bart0sh, 
thank you for the workaround. Besides the CNI configurations under `/etc/cni/net.d`, the extension of the `plugin_dirs` property was missing in the `crio.conf` configuration, because `crio` doesn't seem to look in `/opt/cni` for the rolled out plugins by default. I created a drop-in file.

```bash
$ cat /etc/crio/crio.conf.d/00-plugin-dir.conf
[crio.network]
plugin_dirs = [
  "/opt/cni/bin/",
  "/usr/lib/cni",
]
```

In the Wiki of Arch Linux are booth the problems [documented](https://wiki.archlinux.org/title/CRI-O#Configuration)  that drew my attention to this. 

Thanks for the tip. I would be happy if this could be made a bit more user-friendly, because debugging the error was not easy for me and is probably a big hurdle for kubernetes beginners doing a vanilla installation.

**Comment by bart0sh** [comment]
@volker-raschek
> because crio doesn't seem to look in /opt/cni for the rolled out plugins by default

That sounds quite strange to me. If cri-o doesn't have access to CNI binaries how did you manage to get bridge CNI IPs assigned to your coredns pods?

> I would be happy if this could be made a bit more user-friendly

We're working on this :) This issue is part of the work.
If you want to help you can update [this](https://kubernetes.io/docs/setup/production-environment/container-runtimes/) and [this](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) documentation.

**Comment by bart0sh** [comment]
@haircommander JFYI: podman seems to enable bridge CNI on Ubuntu 20.04
```
$ dpkg -L podman |grep cni
/etc/cni
/etc/cni/net.d
/etc/cni/net.d/87-podman-bridge.conflist
ed@devel ~ $ sudo apt policy podman
podman:
  Installed: 100:3.0.1-2
  Candidate: 100:3.0.1-2
  Version table:
 *** 100:3.0.1-2 500
        500 https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_18.04  Packages
        100 /var/lib/dpkg/status
```

**Comment by matthias50** [comment]
> Is that acceptable?

I think so. To summarize, the proposal is to separate the plugin configuration files: /etc/cni/net.d/100-crio-bridge.conf and /etc/cni/net.d/200-loopback.conf into a separate package, say `containernetworking-plugins` and have this package be "Suggests". This will have the effect of the files not being installed automatically on RPM distros when one does a `dnf install cri-o` Correct?

**Comment by haircommander** [comment]
> Correct?

correct!

> @haircommander JFYI: podman seems to enable bridge CNI on Ubuntu 18.04

I believe, though without any evidence at present, that cri-o should ignore the podman ones, and vice versa

> Besides the CNI configurations under /etc/cni/net.d, the extension of the plugin_dirs property was missing in the crio.conf configuration, because crio doesn't seem to look in /opt/cni for the rolled out plugins by default. I created a drop-in file.

ah, we did this to be able o package the cni plugins without interfering with docker/containerd installation I believe. I think we can extend the list of plugin dirs to both in 1.24 as well

**Comment by bart0sh** [comment]
> I believe, though without any evidence at present, that cri-o should ignore the podman ones, and vice versa

I'm more concerned about Kubernetes here. It will not ignore either of those.

**Comment by willzhang** [comment]
Refer to this link. i want know Is this the final interim solution？
[](https://wiki.archlinux.org/title/CRI-O)

Warning: The cri-o package installs the 10-crio-bridge.conf and 99-loopback.conf examples to /etc/cni/net.d by default . This may conflict with Kubernetes cluster network fabrics (weave, flannel, calico, etc) and require manual deletion to resolve this

**Comment by haircommander** [comment]
I don't think centos 7 can support weak dependencies in an RPM. is it satisfactory to update to weak dependencies (Suggests: containernetworking-plugins) just for centos 8 and above? I think suggests best describes the relationship. I suppose we could also just drop entirely

**Comment by haircommander** [comment]
I have implemented ^ in 1.24.0. PTAL!

**Comment by matthias50** [comment]
> is it satisfactory to update to weak dependencies (Suggests: containernetworking-plugins) just for centos 8 and above?

We are not using CentOS 7 any more so 8 and up is fine from my end.

**Comment by matte21** [comment]
> I have implemented ^ in 1.24.0. PTAL!

I use CRI-O 1.28 on Ubuntu 22.04 yet I'm still facing this issue.

The installation command I'm using is 

>         apt-get install -y cri-o=1.28.1~0 cri-o-runc=1.0.1~2 cri-tools=1.26.0-00

after having configured apt repos and keys, etc...

How can I fix this?

**Comment by kwilczynski** [comment]
@matte21, what issue are you experiencing exactly?

Just to let you know, you replied to an old closed issue about problems with CNI set up on an old CentOS version.
